### PR TITLE
Refactor repeat timer

### PR DESCRIPTION
**Pull request by ethanfrey**
Same cleanup (remove mutex, use command channel) as with throttle timer.
Made the output non-blocking as per time.Ticker
Made Stop idemnepotent

### CODE REVIEW

File: common/repeat_timer.go [hunk 1]
```diff
@@ -11,76 +10,93 @@ It's good for keeping connections alive.
...
-	ticker *time.Ticker
-	quit   chan struct{}
-	wg     *sync.WaitGroup
-	dur    time.Duration
+	dur     time.Duration
+	ticker  *time.Ticker
+	stopped bool
 }
 
+type repeatCommand int32
```
Comment (melekes) [inline common/repeat_timer.go]: why int32?) int8 would cut it, no

File: common/repeat_timer.go [hunk 2]
```diff
@@ -11,76 +10,93 @@ It's good for keeping connections alive.
...
-		close(t.quit)
-		t.wg.Wait() // must wait for quit to close else we race Reset
-		t.ticker = nil
 	}
-	return exists
+}
+
+// trySend performs non-blocking send on t.Ch
+func (t *RepeatTimer) trySend() {
+	// NOTE: this was blocking in previous version (t.Ch <- t_)
```
Comment (melekes) [inline common/repeat_timer.go]: we'll need to test Tendermint (or where this timer is used) if we're gonna change behaviour

File: common/repeat_timer.go (hunk 1 above)
Comment (ethanfrey) [reply common/repeat_timer.go]: yup

File: common/repeat_timer.go (hunk 2 above)
Comment (ethanfrey) [reply common/repeat_timer.go]: will return to old behavior and leave comment for now

**Comment by ethanfrey** [comment]
Okay, anton, take another look if you want

**Comment by ethanfrey** [comment]
Thanks for the cleanup and merge!

File: common/repeat_timer.go [hunk 3]
```diff
@@ -11,76 +10,93 @@ It's good for keeping connections alive.
...
 		}
-		close(t.quit)
-		t.wg.Wait() // must wait for quit to close else we race Reset
-		t.ticker = nil
 	}
-	return exists
+}
+
+// send performs blocking send on t.Ch
+func (t *RepeatTimer) send(tick time.Time) {
```
Comment (ebuchman) [inline common/repeat_timer.go]: we should defntly not block! I think this is responsible for some of the recent non-deterministic failures on Tendermint. I'll change this

**Pull request merged**